.github/workflows/ci.yml

```diff
@@ -228,39 +151,18 @@ jobs:
           echo "PATH=/usr/lib/ccache:$PATH" >> $GITHUB_ENV
           echo -e "127.0.0.1\t$HOSTNAME" | sudo tee -a /etc/hosts > /dev/null
       - name: build
-        run: ./bootstrap.sh -e -t && cd build  && cmake -DUSE_SHARED_LIBS= -DCMAKE_BUILD_TYPE=Release -DENABLE_NANOFI=ON -DENABLE_JNI=ON -DENABLE_SENSORS=ON -DENABLE_OPENWSMAN=ON -DENABLE_OPENCV=ON -DENABLE_MQTT=ON -DENABLE_GPS=ON -DENABLE_USB_CAMERA=ON -DENABLE_LIBRDKAFKA=ON -DENABLE_OPC=ON -DENABLE_SFTP=ON -DENABLE_COAP=ON -DENABLE_PYTHON=ON -DENABLE_SQL=ON -DENABLE_AWS=ON -DSTRICT_GSL_CHECKS=AUDIT -DFAIL_ON_WARNINGS=ON -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ .. &&  cmake --build . --parallel 4
-      - name: test
-        run: cd build && make test ARGS="--timeout 300 -j8 --output-on-failure"
-  ubuntu_16_04_all:
-    name: "ubuntu-16.04-all"
-    runs-on: ubuntu-16.04
-    timeout-minutes: 90
-    steps:
-      - id: checkout
-        uses: actions/checkout@v2
-      - id: cache
-        uses: actions/cache@v2
-        with:
-          path: ~/.ccache
-          key: ubuntu-16.04-all-ccache-${{github.ref}}-${{github.sha}}
-          restore-keys: |
-            ubuntu-16.04-all-ccache-${{github.ref}}-
-            ubuntu-16.04-all-ccache-refs/heads/main-
-      - id: install_deps
         run: |
-          sudo apt-add-repository -y "ppa:ubuntu-toolchain-r/test"
-          sudo apt update
-          sudo apt install -y ccache openjdk-8-jdk maven libusb-1.0-0-dev libpng12-dev libgps-dev libsqliteodbc
-          sudo ln -s /usr/lib/x86_64-linux-gnu/odbc/libsqlite3odbc.so /usr/lib/x86_64-linux-gnu/libsqlite3odbc.so
-          echo "PATH=/usr/lib/ccache:$PATH" >> $GITHUB_ENV
-          echo -e "127.0.0.1\t$HOSTNAME" | sudo tee -a /etc/hosts > /dev/null
-      - name: build
-        run: sudo mount tmpfs -t tmpfs /tmp && ./bootstrap.sh -e -t && cd build  && cmake -DUSE_SHARED_LIBS= -DENABLE_NANOFI=ON -DENABLE_OPENWSMAN=ON -DENABLE_OPENCV=ON -DENABLE_GPS=ON -DENABLE_USB_CAMERA=ON -DENABLE_LIBRDKAFKA=ON -DENABLE_OPC=ON -DENABLE_SFTP=ON -DENABLE_MQTT=ON -DENABLE_COAP=ON -DENABLE_PYTHON=ON -DENABLE_SQL=ON -DENABLE_AWS=ON -DENABLE_AZURE=ON -DSTRICT_GSL_CHECKS=AUDIT -DFAIL_ON_WARNINGS=ON .. &&  cmake --build . --parallel 4
+          ./bootstrap.sh -e -t
+          cd build
+          export CC=clang
+          export CXX=clang++
+          cmake -DUSE_SHARED_LIBS= -DCMAKE_BUILD_TYPE=Release -DENABLE_NANOFI=ON -DENABLE_JNI=ON -DENABLE_SENSORS=ON -DENABLE_OPENWSMAN=ON -DENABLE_OPENCV=ON -DENABLE_MQTT=ON -DENABLE_GPS=ON -DENABLE_USB_CAMERA=ON -DENABLE_LIBRDKAFKA=ON -DENABLE_OPC=ON -DENABLE_SFTP=ON -DENABLE_COAP=ON -DENABLE_PYTHON=ON -DENABLE_SQL=ON -DENABLE_AWS=ON -DSTRICT_GSL_CHECKS=AUDIT -DFAIL_ON_WARNINGS=ON ..
+          cmake --build . --parallel $(($(nproc) * 2))
```

Comment:
What the reason for $(nproc) * 2 instead of just $(nproc)?

Reply:
It might improve CPU utilization as not all make threads use the CPU all the time. For example, when one thread is downloading a tarball, another compilation thread could still use that core, as IO doesn't need too much cpu. And even when all threads are compiler/linker threads, fully utilizing the CPU, the performance degradation is negligible compared to doing the same over $(nproc) threads.
A drawback of going higher is that more threads need more memory, and if we run out, swapping kills compilation times. Linking is especially memory-heavy. On Gentoo, 32C/64T 128G RAM, when I run 4 portage compilation tasks on 32 threads each, I tend to test the limits of even 128 GB of RAM, and some cores are still idling the majority of the time, even after going 2x the threads.
The general advice from a few years ago was ranging from N, N+1, N+2 to 2*N make jobs per cpu thread.

Reply:
I can revert to $(nproc) if you prefer it that way. The main motivation was getting rid of hardcoded core numbers to be able to utilize more powerful executors, and I thought if I'm already changing it, might as well make sure that we maximize CPU utilization.

Reply:
I support removing the hardcoded core numbers, I wasn't sure if the build can utilize core numbers above $(nproc). If it's okay and we do not run into any problems I'm okay with keeping the $(nproc) * 2 job count.

Reply:
I reverted to $(nproc), because we already ran into problems: the ubuntu 20.04 job timed out. I think it's more likely to be because of the number of extensions and the fact that these executors are not great, but it's worth a try.